### ISSUE TITLE
[bug] PST Many spell effects are excessively delayed

### ISSUE THREAD

**Issue by chilvence** [gemrb/gemrb#637]
I have been slowly working through testing these the last few weekends to finish #587 , and if I had to choose one problem in the code that basically ruins everything, it is definitely this one

Many Spells that would otherwise seemingly work fine suffer from uneeded delays before taking effect. If you cast the spell spwi111 ‘shield’ for example, you can be waiting several rounds for it to actually visibly appear.

- Essentially, this problem seems to be in every spell or effect that uses timing mode 3 or 4  delay - limited / delay - permanent 

- In Torment.exe The ‘duration’ in the effects seems to control the delay before they should appear, to allow timing with the actor spellcasting animation, however Torment seems to measure this in frames or AI updates, while GemRB seems to count out whole seconds or rounds before displaying the bam effect. This can be seen by modifying the effect duration of spwi111 (shield) to 1, after which it will take effect with much more sensible timing. 

- To me that says that GemRB is just using the wrong measure of time when applying these effects to the target actor, but I don't know where I should be looking to figure that one out. 

 - Many of them also seem to have the c9 [201] “play bam with fx” effect with duration set like ‘15’ or ‘30’, but for these spells the actual duration of the spell effects are over-ridden by this PST opcode, because unlike BG/IWD they only have one ‘level’ stored in the spl file  : 

https://github.com/gemrb/gemrb/blob/00fac6b7b0ccff21820d50d0d5029b54f30fe435/gemrb/plugins/PSTOpcodes/PSTOpcodes.cpp#L575

The ‘Converttiming’ function seems to be changing the timing mode of the effects, which only adds to my confusion: https://github.com/gemrb/gemrb/blob/00fac6b7b0ccff21820d50d0d5029b54f30fe435/gemrb/plugins/PSTOpcodes/PSTOpcodes.cpp#L570 

This may or may not interfere with the timing, it possibly doesn’t but I thought it might be worth an eyeball just in case

- Spells affected:
SPWI111 Shield			(anim and effects delayed)
SPWI114 Friends			(effects delayed)
SPWI119 Submerge the will		(effects delayed)
SPWI202 Black Barbed Curse	(whole spell mistimed)
SPWI204 Blood Bridge		(anim delayed)
SPWI211 Strength			(anim delayed)
SPWI213 Horror			(sound out of sync with anim)
SPWI214 Luck			(sound and effects delayed)
SPWI215 Knock			(sound played long after effect/anim)
SPWI217 Ignus’ Terror		(same as horror)
SPWI219 Power of one		(anim and sound delayed)
SPWI304 Cloak of warding		(anim and effects delayed)
SPWI308 Tasha’s Laughter		(entire spell delayed)
SPWI309 Vampiric Touch		(unable to confirm spell works)
SPWI311 Zerthimon’s Focus		(anim and effects delayed)
SPWI312 Infernal shield		(anim and effects delayed)
SPWI402 Confusion			(amin and effects delayed)
SPWI409 Remove Curse		(whole spell delayed)
SPWI502 Cone of Cold		(certain effects delayed)
SPWI505 Desert Hell			(whole spell delayed. also, only partially implemented?)
SPWI704 Guardian mantle		(anim and effects delayed)
note: most of the high level spells need special work anyway so I stopped testing about here
Some priest spells are also affected unsurprisingly
SPPR106 Protection from evil	(anim and effects delayed)
SPPR302 Flamewalk			(anim and effects delayed)
ETC...

It is hard to definitively say whether all of these spells are otherwise working correctly, but for nearly all of them, as long as I waited patiently for up to a minute or so, many of their effects seemed to play out as they do in the original game (aside from minor gfx differences here and there)

There are actually a lot more spells already at least partially implemented than I originally thought as well, which was a nice surprise.

**Comment by lynxlynxlynx** [comment]
Thanks a bunch! 

When you mention fx_overlay with duration of 15/30, that would be in seconds, right (3/6 rounds; if ticks: 1/2 seconds)? Does it use a different timing mode, not one of these problematic delayed ones?

**Comment by chilvence** [comment]
I can't honestly suggest what exactly the timing should be, because I just don't know the infinity engine that well (even after years playing I should, but...). 

I think it might be the 'ai ticks' option, as it scales even in the original executable when I adjust that via the "Maximum Frame Rate=___" option in torment.ini, if I set it to something silly like '10' it takes proportionally longer to play the animations.

It definitely isn't meant to be seconds, that would be much too long. Just taking the shield spell as an example, the graphical and sound effect for the shield animation are both timed as 15, and, are supposed to appear  almost the instant the spell is cast, so whatever it is supposed to be, it is leaning toward fractions of a second rather than multiples.

**Comment by lynxlynxlynx** [comment]
Ok. Do you think it's sensible to continue with #587 until this is fixed? Just trying to manage the priorities somehow.

**Comment by chilvence** [comment]
Well, I already have a lot more notes to add regarding eliminating that list, but I still haven't fully fleshed them out. There are about five or six common issues, but I would rather test things from every angle I can before I report them.

This one particular issue does make it nearly impossible to test and verify all the affected spells however, and unlike some of the other issues it does also have a profoundly negative affect on game playability if you are a mage, if I was confident enough to think I could fix it I would definitely be spending my free time on that instead of just testing .

My most recent game I played as a mage (I have always tended to just thug my way through the game when I played it before, since the magic has always been a bit shallow compared to BG2), it has been a bit gruelling from start to finish because of this one problem. 

But on the other hand, the actual game itself is in so much better shape compared to the first time I ever played it, it finally feels possible to pin down the smaller details.

**Comment by burner1024** [comment]
Probably related: #144

**Comment by lynxlynxlynx** [comment]
Not really, this is about the modified effects system pst used.

**Comment by chilvence** [comment]
I may have finally figured out one aspect of this (It took me quite a while, but a lot of that was just reading the source and trying to learn what it is doing)

However, I did manage to get some of the spells to work as expected, by changing very little! It was in the end all about the multiplier factor that is applied to the initial duration when the spell effects are applied. 

I need to test everything all over again because the quick change I made has probably broken other spells, but I am getting closer to finally solving this one!

**Comment by chilvence** [comment]
The only description I have ever found on how this is supposed to work was made by [Galactygon](https://forums.beamdog.com/discussion/comment/866592/#Comment_866592) over at the Beamdog forums 

>  While timing mode duration uses seconds (just like the other IE games), the timing mode delay and delayed-duration uses ticks which allow you to fine-tune when graphics/sound effects play

It confirms among already obvious things, that the 'duration' field in the data from of all these spells is supposed to be measured in ticks instead of seconds. 

I tested this out by modifying the 'PrepareDuration" macro in EffectQueue.h to not multiply by AI_UPDATE_TIME, and that was all it took to make basically all of the spells _appear_ to cast correctly. I put the emphasis on "appear", because that also means that they almost instantly expire, but that is not because the previous behaviour is right, is is because the actual duration was always calculated by the original EXE. 

> Many of the spells use hardcoded projectiles just like the original PS:T. _**You might be wondering why some spells such as "Strength" or "Power of One" do not have any non-graphical effects in their extended headers.**_ That's because the projectile calculates durations and stat changes just like the original game. Many other projectiles calculate level-based duration (i.e. Tasha's Unbearable Derisive Laughter) where it appears information is missing in the extended header compared to BG(2)(EE). The code has been simply ported over from the original PS:T.

The 'duration' of these spells was only used for fine tuning the timing of casting animations, and everything else was done in the actual code. This is why they don't have multiple effect headers , and the only reason they seem to expire prematurely, or last forever. All the data is a red herring, it would have been totally ignored.

In other words, the actual duration that the effects of these spells are supposed to have has nothing to do with the .SPL file, that simply must be calculated by the core "effect" code. The only important piece of information used is the effect index.

With that knowledge asserted, the only thing that needs to change is the calculation of the longevity of the effects, which would be based on AD&D rules, pretty much boils down to "Duration = X+1 round per caster level", which even if not exact, would still be a drastic improvement. ( the quest to find my lost 2e manual is basically the last part of this )

**Comment by lynxlynxlynx** [comment]
Ok, so it was as we feared. :( Like you said in the first post, it appears the duration field for delayed timing modes is just used for the delay, but in ticks, while the duration is always recalculated. 

The manual is useless, but the ingame descriptions should be fine as far as duration specs go.

Looks like #458 and  #641 might be duplicates, which is good. :)

**Comment by lynxlynxlynx** [comment]
Looking at shield, it uses:
- fx_play_bam_blended, delay/permanent
- fx_overlay, delay/permanent
- sound effect, irrelevant

The overlay opcode is the main culprit and it already converts timing. It needs to be reviewed if it was reverse engineered properly. Perhaps the delay timing can be fixed there as well.

**Comment by chilvence** [comment]
I didn't get much further than just isolating the locations of the opcodes when I tried with ida, but I think an educated guess based on the descriptions would be fine for most cases. I still haven't got as far just 'playing the game' yet, but I'll be experimenting with that angle now I know where to look. Something like 'Caster Level x AI_UPDATE_TIME' should be enough to add the per-level duration bonus, I am just not sure about the base durations.

**Comment by lynxlynxlynx** [comment]
I think you're complicating the matter. Testing shield, it appears to work fine, just the onset delay is the problem. So nothing with the actual opcodes. This fixes it for me, try it with other spells:
http://sprunge.us/9JPgzk?diff

**Comment by chilvence** [comment]
Probably true, I usually approach the code more or less like a car mechanic would approach surgery :) That already looks like the change I roughly tested, I just hadn't got to isolating it to just the case where it is needed.

Soon as I finish with today's regular events though, I'll get back to carefully comparing the spells and noting the results
